### ISSUE TITLE
Multiple parallel subscribe actions sometimes crashing app

### ISSUE THREAD

**Issue by dnmsk** [burakoner/ApiSharp#1]
I had this exception from Bybit.Api library with small improvements to use latest ApiSharp. 
```
Unhandled exception. System.AggregateException: One or more errors occurred. (Collection was modified; enumeration operation may not execute.)
 ---> System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.Queue`1.Enumerator.MoveNext()
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source, Int32& length)
   at ApiSharp.Models.AsyncResetEvent.<>c__DisplayClass5_0.<WaitAsync>b__0() in E:\Github\ApiSharp\ApiSharp\Models\AsyncResetEvent.cs:line 51
   at System.Threading.CancellationTokenSource.Invoke(Delegate d, Object state, CancellationTokenSource source)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
   --- End of inner exception stack trace ---
   at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
[1]    174402 IOT instruction (core dumped)
```

Could be using  simple `_lockObject = new object();` for synchronization in AsyncResetEvent?

**Comment by burakoner** [comment]
I think it has been fixed in ApiSharp 2.0.0. Please try ByBit.Api 1.2.0. If you have trouble again please let me know

**Comment by dnmsk** [comment]
Hi! 2.0.3. Same issue. Crushing Bybit and Okx.
I think there could be different lock object
![image](https://github.com/burakoner/ApiSharp/assets/12471602/f3f57513-6841-45f2-baa4-ed3c0d9d02f0)

**Comment by burakoner** [comment]
inspecting deeply

**Comment by burakoner** [comment]
can u share me your codes please

**Comment by dnmsk** [comment]
Yep, sure. There is https://github.com/dnmsk/ApiSharpApiExample
Launch and wait it for some time

**Comment by dnmsk** [comment]
I've no idea what happening in that code, but think that follow implementation won't throw concurrent dependent exception

```
namespace ApiSharp.Models;

/// <summary>
/// Async auto reset based on Stephen Toub`s implementation
/// https://devblogs.microsoft.com/pfxteam/building-async-coordination-primitives-part-2-asyncautoresetevent/
/// </summary>
public class AsyncResetEvent : IDisposable
{
    private static readonly Task<bool> _completed = Task.FromResult(true);
    private readonly List<TaskCompletionSource<bool>> _waits = new();
    //private Queue<TaskCompletionSource<bool>> _waits = new();
    private bool _signaled;
    private readonly bool _reset;

    /// <summary>
    /// New AsyncResetEvent
    /// </summary>
    /// <param name="initialState"></param>
    /// <param name="reset"></param>
    public AsyncResetEvent(bool initialState = false, bool reset = true)
    {
        _signaled = initialState;
        _reset = reset;
    }

    /// <summary>
    /// Wait for the AsyncResetEvent to be set
    /// </summary>
    /// <param name="timeout"></param>
    /// <returns></returns>
    public Task<bool> WaitAsync(TimeSpan? timeout = null)
    {
        lock (_waits)
        {
            if (_signaled)
            {
                if (_reset)
                    _signaled = false;
                return _completed;
            }
            else
            {
                var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
                if (timeout != null)
                {
                    var cancellationSource = new CancellationTokenSource(timeout.Value);
                    var registration = cancellationSource.Token.Register(() =>
                    {
                        tcs.TrySetResult(false);

                        // Not the cleanest but it works
                        lock (_waits)
                        {
                            var ourWait = _waits.FirstOrDefault(w => w == tcs);
                            if (ourWait != null)
                            {
                                _waits.Remove(ourWait);
                            }
                        }
                    }, useSynchronizationContext: false);
                }

                _waits.Add(tcs);
                return tcs.Task;
            }
        }
    }

    /// <summary>
    /// Signal a waiter
    /// </summary>
    public void Set()
    {
        lock (_waits)
        {
            if (!_reset)
            {
                // Act as ManualResetEvent. Once set keep it signaled and signal everyone who is waiting
                _signaled = true;
                for (var i = 0; i < _waits.Count; i++)
                {
                    _waits[i].TrySetResult(true);
                }
                _waits.Clear();
            }
            else
            {
                // Act as AutoResetEvent. When set signal 1 waiter
                if (_waits.Count > 0)
                {
                    for (var i = 0; i < _waits.Count; i++)
                    {
                        _waits[i].TrySetResult(true);
                    }
                    _waits.Clear();
                }
                else if (!_signaled)
                    _signaled = true;
            }
        }
    }

    /// <summary>
    /// Dispose
    /// </summary>
    public void Dispose()
    {
        lock (_waits)
        {
            _waits.Clear();
        }
    }
}

```

**Comment by burakoner** [comment]
I'm in hospital and cant write code or do any test. I will ve able to check it on monday. In my mind

**Comment by dreamzgj** [comment]
var registration = cancellationSource.Token.Register(() =>
                    {
                        tcs.TrySetResult(false);

                        List<TaskCompletionSource<bool>> _tmp_waits = _waits.ToList();
                        // Not the cleanest but it works
                        _waits = new Queue<TaskCompletionSource<bool>>(_tmp_waits.Where(i => i != tcs));
                    }, useSynchronizationContext: true);  //here !!! use true can fix

**Comment by burakoner** [comment]
Queue or List is not thread safe structures. I've fixed it with ConcurrentQueue. I tested with your codes again and got no error this time. I'll publish it tonight. Thanks for your feedback and patience

**Comment by burakoner** [comment]
Fixed in Version 2.0.5
